### ISSUE TITLE
Microsoft Login

### ISSUE THREAD

**Issue by failip** [SKCraft/Launcher#376]
A way of logging in via the Microsoft Account should be added, since new Accounts have to be linked to Microsoft Accounts now and all Accounts have to be converted to Microsoft Accounts sometime next year.  The Microsoft Login is a simple OAuth 2.0 Login as described [here](https://wiki.vg/Microsoft_Authentication_Scheme). 

I tried Implementing it but the first step has to be done in a Browser/WebView which I currently can't see a way to do with Java Swing only. The later parts of the Authentication should be straightforward.

Links:
- https://wiki.vg/Microsoft_Authentication_Scheme
- https://github.com/MiniDigger/MiniLauncher/blob/master/launcher/src/main/java/me/minidigger/minecraftlauncher/launcher/gui/MsaFragmentController.java

**Comment by hedgehog1029** [comment]
This is being worked on.

**Comment by majorcyto** [comment]
I second this please, I have some friends who only got Minecraft recently so they won't be able to play with us until Microsoft Login is also supported. 😢

**Comment by spannerman79** [comment]
@hedgehog1029 be **extremely** careful if you decide to use the data over on https://wiki.vg/Microsoft_Authentication_Scheme as it isn't an official solution.  Other launchers that use that can be issued with a C&D or even a take down notice as its using the API for not what it was intended.

A Microsoft Azure application is inteded to be used on the Azure platform. ie cloud/SaaS.

A dev for another launcher (MultiMC) actually works over at Mojang/MS on the vanilla launcher and its tooling ( [Source](https://twitter.com/peterixxx/status/996105344584765442) ) and he refuses to impliment https://wiki.vg/Microsoft_Authentication_Scheme into MultiMMC due to the legal mindfield that it will create.

**Comment by failip** [comment]
[Here](https://github.com/MultiMC/MultiMC5/issues/3392#issuecomment-713796443) he said he's working on an official solution. Guess this will have to wait until that's implemented.

**Comment by hedgehog1029** [comment]
I'm relatively certain I am not going to be issued a copyright takedown notice for implementing an OAuth pipeline.

From some initial tests, Azure apps actually won't let you convert a Microsoft token to an Xbox token. So the solution is a little more complex that I'd hoped.

**Comment by Murachacha** [comment]
Hey, any updates on this?

**Comment by jkroepke** [comment]
There is a library from Microsoft called https://github.com/AzureAD/microsoft-authentication-library-for-java

It might be easier instant implement it on your own.

**Comment by Nydid** [comment]
> Hey, any updates on this?

Pretty much waiting on Microsoft and MultiMC devs

**Comment by MadMartian** [comment]
> [Here](https://github.com/MultiMC/MultiMC5/issues/3392#issuecomment-713796443) he said he's working on an official solution. Guess this will have to wait until that's implemented.

The solution might still be difficult to implement, federated identity provision for non-web clients has headaches particularly with an SPA login provider.  I do this professionally during the day and that's why we expose private / internal APIs that clients can consume to generate JWT links (action tokens) to bypass convoluted web flows.  I don't know how you would do that securely in the DMZ though.

**Comment by jkroepke** [comment]
> to bypass convoluted web flows.
This is compatible with the Microsoft Account MFA?

I guess the cleanest way is to open a Browser. The browser could be the system browser, the does not need to implement a own browser.

Additionally, headless clients could just use https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code

**Comment by TheVegen** [comment]
The SKCraft launcher has worked great for us and it is very concerning that our mod will stop working when all the accounts get kicked over to being Microsoft accounts. Any idea when a fix/workaround will be found for this?

**Comment by hedgehog1029** [comment]
there are plenty of non-code grant flows available if you're doing Azure auth - but it seems that you can't turn an arbitrary app token (e.g. custom Azure app auth) into Xbox tokens, you just get an unhelpful error message.

The OpenXbox project seems to have concluded that in order to do Xbox auth you need to use a special client ID, which is effectively "the xbox id" - and only the primary oauth flow works for that.

I need to ask around the mcdevs channels about who wrote that wiki.vg article, and whether they actually tested custom azure apps.

**Comment by jkroepke** [comment]
Maybe adopt a [solution](https://github.com/MiniDigger/MiniLauncher/blob/master/launcher/src/main/java/me/minidigger/minecraftlauncher/launcher/gui/MsaFragmentController.java) from an other launcher is a good intermediate solution unless something [offical](https://github.com/MultiMC/MultiMC5/issues/3392#issuecomment-713796443) is available.

**Comment by sylphio** [comment]
AtLauncher implementation as reference : https://github.com/ATLauncher/ATLauncher/issues/415

**Comment by johnfriedrich** [comment]
I don't see a reason for an azure app at all. Why not implementing a standard o auth solution within a web view? Thats secure (enough) aswell and is a common standard

**Comment by hedgehog1029** [comment]
good news: custom Azure OAuth apps actually *do* work (i had a horrible, horrible 1-character typo which was causing the XBL request to fail), which means we can do both "launch the system browser" and device code based auth.

this means I can implement this without having to rely on a webview! which is nice and means I don't have to install JavaFX

**Comment by hedgehog1029** [comment]
Alright, experimental-ish implementation is up on branch [`feat/new-authentication`](https://github.com/SKCraft/Launcher/tree/feat/new-authentication).

I don't actually have a valid Microsoft Minecraft account to test with, only an Xbox account, so it's tested only up to a certain point. Let me know if something breaks.

**Comment by sylphio** [comment]
i did a test.

https://pastebin.com/Ua5XfR1f

The login part work well, but it crash once  SKCraft is authorized into the account.

**Comment by hedgehog1029** [comment]
> i did a test.
> 
> https://pastebin.com/Ua5XfR1f
> 
> The login part work well, but it crash once SKCraft is authorized into the account.

That's a reasonably expected error, except I didn't have any special handling for it. I've just pushed 1d4e46214ba0288fb5246cd83c2d96399a9bc9cc which handles these errors and presents a more human-readable error code :)

**Comment by sylphio** [comment]
Hi, 

Sadly the %d is not expanded and so it only show Err: %d.

**Comment by hedgehog1029** [comment]
Fixed the bad expansion, that was me being silly.

That does mean it was an unknown error of some kind, would be interesting to know what the code is though.
I'm not sure what's up with account switching, it seems Microsoft's OAuth likes to pick an account and stick with it?

**Comment by sylphio** [comment]
What i got is   (Xerr 0)

FYI, in the browser windows everything look fine. I get the "OK: you can close the browser now".
What browser are you using for your test? Chrome is my default browser.

**Comment by hedgehog1029** [comment]
It sounds like the Microsoft auth leg is working fine, but it errors when authorizing with XSTS (the second leg of Xbox authentication) - that's why you get an XErr code.

Weird that you get a code 0. I suppose I should check for that and see if the error message is non-empty.

**Comment by hedgehog1029** [comment]
I should probably have been clearer, lol: it sounds like the browser leg is working fine (you get the OK and you get an Xbox error). Chrome is fine, I'm sure.

I pushed a quick fix that will display any returned error message if it's non-empty and not an otherwise known code. I have no idea if your code 0 response does contain a non-empty message, though.

**Comment by sylphio** [comment]
Just tested but nothing really interesting in the log. 

https://pastebin.com/sgTxykPu

Same message box as before.  A unknown error    (stuff)    Xbox (XErrr 0)

**Comment by hedgehog1029** [comment]
Yeah, sounds like an empty message. Not particularly helpful.

@sylphio Are you sure you're logging in with the correct account? That's the only thing I can think of - I can get further (into the Minecraft auth leg) before I get an error because my Microsoft account doesn't own Minecraft, but does have an Xbox account. So the fact you're failing at the XSTS step with a useless error code is weird to me.

Everyone else - has anyone successfully logged in with a Microsoft account on this branch? Hearing a success story would be encouraging :)

**Comment by sylphio** [comment]
Okay,  i took a look with a debugger this time. 

Here is the full error
XstsError(xErr=0, message=, redirect=https://start.ui.xboxlive.com/AddChildToFamily)

The account i was using to test is the brand new one of my daughter. 
I think the xsts was simply redirecting the login so i can add her to my "Familly" first but the redirect is ignored  by skcraft now.

So i added her to my familly and now it work ;p

Thank :)

**Comment by hedgehog1029** [comment]
oh weird so maybe the XErr reading is broken. i will have to see if I can sign in with a non-xbox account!

the "redirect" provided by the error response doesn't resolve on desktop, it's likely meant to be a shortcut for Xbox consoles.

**Comment by hedgehog1029** [comment]
Given that we now have a successful login - I will be fixing up lingering bugs & looking to get this merged today.

**Comment by sylphio** [comment]
After the initial login,  the account is correctly added to the account list but after i restart the launcher it crash on "relog"

https://pastebin.com/jNQyJ9tp

**Comment by hedgehog1029** [comment]
> After the initial login, the account is correctly added to the account list but after i restart the launcher it crash on "relog"
> 
> https://pastebin.com/jNQyJ9tp

oh WHOOPS I can see exactly where the problem is for that - will fix that in today's batch

**Comment by hedgehog1029** [comment]
Okay, pushed fixes for the bugs mentioned above (XErr is deserialized correctly now, & relogin should work) and also made Microsoft's OAuth actually show an account selector, which is useful!

@sylphio let me know if the relogin process works on the new build, cheers for all your testing help :)

**Comment by hedgehog1029** [comment]
I think we're getting there at least! Having the game launched is good progress.

Do you see your Minecraft skin face show up in the SKCraft account selector? I'm wondering if the profile response went weird and the UUID is somehow missing...

**Comment by sylphio** [comment]
The account name is fine but it display the default skin and not the good one.

**Comment by hedgehog1029** [comment]
That does kind of imply to me that something is going wrong with the UUID. Though I would expect it to throw an error if fetching the skin fails. So that's kind of weird.

I suppose it's possible that the profile response is returning a "null UUID" e.g. all zeroes. That would return a Steve skin face but be invalid for playing on servers. You might be able to see this if you look in the launcher console (options -> console in the lower left) as all the game arguments are printed when launching the game.

**Comment by sylphio** [comment]
Nope the UID is valid and is the good one :/

https://pastebin.com/2uUtDtTq

**Comment by sylphio** [comment]
I compared the log from one launch with a mojang account and one with microsoft account and they both look fine. The access token is far longer in M$ authentication.

**Comment by hedgehog1029** [comment]
Yeah I think I'm in agreement that the UUID is fine - Visage seems to be returning a Steve face for your UUID at the moment, which is weird but it'll probably uncache at some point and you can remove & readd the account to get the correct skin.

I had a quick check and I'm definitely returning the correct token in the session instance. So I assume the access token being longer is just a side effect of the new authentication system.

Just as a test: singleplayer works? You can F5 to see your skin & typing in chat works? I'm pretty sure you are authenticated correctly on the client but would be good to actually know the session is working.

**Comment by sylphio** [comment]
Yep single player/chat work and the skin display properly.

**Comment by sylphio** [comment]
FYI i compared the launch line between SKCraft and Atlauncher. 
Access token is same lenght BUT!
ATLauncher use --userType Xbox and not --userType, microsoft

**Comment by hedgehog1029** [comment]
From what I can see, even 1.16 doesn't recognize the new user type, but it will just assign null into a field and that field has zero accessors anyway, so I don't think it matters. Especially for 1.7.10 which is loooong before microsoft auth obviously.

Do things work under ATLauncher (under 1.7.10)? Because that's more interesting/concerning, heh.

**Comment by sylphio** [comment]
i quickly moved my modpack in a atlaucher directoryto test out  and i got the same error that with SkCraft.

failed to login : null

Do you know if it require some change server side?

**Comment by hedgehog1029** [comment]
I think this is drifting into a separate issue so I've started discussion #397 for it.

**Comment by hedgehog1029** [comment]
Alright, we got that discussion resolved and demonstrated a successful full login & connection to multiplayer servers. So I will just be cleaning up any remaining bits & pieces before getting this merged.

**Comment by spannerman79** [comment]
Well it took MS awhile to update their FAQ...

https://help.minecraft.net/hc/en-us/articles/360050865492-Minecraft-Java-Edition-Account-Migration-FAQ#article-container:~:text=WILL%20THIRD-PARTY%20LAUNCHERS%20CONTINUE%20TO%20WORK%3F,third-party%20launchers%20at%20your%20own%20risk

**Comment by hedgehog1029** [comment]
standard "at your own risk" disclaimer, they don't want to be liable for external security issues

**Comment by Aughen** [comment]
https://www.minecraft.net/en-us/article/last-call-voluntarily-migrate-java-accounts

Coming
Up
Soon :(

**Comment by johnfriedrich** [comment]
> https://www.minecraft.net/en-us/article/last-call-voluntarily-migrate-java-accounts
> 
> Coming Up Soon :(

What do you mean? Support for Microsoft Accs was added months ago

**Comment by MadMartian** [comment]
Yes, we've been using Microsoft Login in our SKLauncher fork without any issues for about a year now.

**Comment by Aughen** [comment]
Just the force migration, that is all :)
